hooks/tk-multi-publish2/ingest/ingest_files.py

```diff
@@ -0,0 +1,198 @@
+
+        return desc + "<br><br>" + """
+        After Ingesting the file a publishedFile gets created. The publishedFile will also be linked to a Plate entity.
+        """
+
+    def publish(self, task_settings, item):
+        """
+        Executes the publish logic for the given item and task_settings.
+
+        :param task_settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the task_settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+        """
+
+        super(IngestFilesPlugin, self).publish(task_settings, item)
+
+        # create a plate entity after the publish has went through successfully.
+        plate_entity = self._create_plate_entity(item)
+
+        # let's create ingest_plate_data within item properties,
+        # so that we can link the version created to plate entity as well.
+        item.properties["ingest_plate_data"] = plate_entity
+
+        # link the publish file to our plate entity.
+        self._link_published_files_to_plate_entity(plate_entity, item)
+        self.logger.info("Plate entity registered and Publish file is linked!")
+
+    def finalize(self, task_settings, item):
+        """
+        Execute the finalization pass. This pass executes once
+        all the publish tasks have completed, and can for example
+        be used to version up files.
+
+        :param task_settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the task_settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+        """
+
+        super(IngestFilesPlugin, self).finalize(task_settings, item)
+
+        # get the data for the plate that was just created in SG
+        plate_data = item.properties["ingest_plate_data"]
+
+        path = item.properties["path"]
+
+        self.logger.info(
+            "Plate created for file: %s" % (path,),
+            extra={
+                "action_show_in_shotgun": {
+                    "label": "Show Plate",
+                    "tooltip": "Open the Publish in Shotgun.",
+                    "entity": plate_data
+                }
+            }
+        )
+
+    @staticmethod
+    def _validate_plate_entity(item):
+        sg = get_sg_connection()
+        sg_filters = [
+            ['project', 'is', item.context.project],
+            ['code', 'is', item.properties["publish_name"]]
+        ]
+        if item.context.entity:
+            if item.context.entity["type"] == "Shot":
+                sg_filters.append(['shots', 'is', item.context.entity])
+            elif item.context.entity["type"] == "Sequence":
+                sg_filters.append(['sg_sequence', 'is', item.context.entity])
+
+        result = sg.find_one(
+            entity_type='Element',
+            filters=sg_filters,
+            fields=['shots', 'code', 'id']
+        )
+        return result
```

Comment:
Perhaps throw an exception if the result is Null?

Reply:
This method is being used in _create_plate_entity, if it doesn't find a plate_entity corresponding to the item. It will create a new entity.

Reply:
The SG python api will at times throw an exception. Wrapping the call to _find_plate_entity() in a try/except, adding your own exception if the result is null, and then handling the exception in _create_plate_entity() is just a more robust solution.

Reply:
If it throws an Exception, should we still proceed by just creating a new plate? Or should we stop the ingestion completely, and raise an error?

Does that also mean all the SG python api can throw exceptions?

Reply:
I would think stop the ingestion for that item and log an error, but don't raise an exception so that the rest can proceed, but I leave it up to you as to how critical you think it should be.

Reply:
Yes, all python api calls can throw exceptions.

Reply:
I can log it as an error and let the other items continue. But, what if the publish entity gets created but the plate entity fails. Do we need to change the validation/publish as well? So that it works with an existing publish file and linking it to the plate? Like the plate_ingest does too.

Reply:
I would actually suggest switching the order so the publishedfile creation occurs after the plate entity creation, so that you can handle cleaning up the plate entity if the publishedfile entity creation fails.

Reply:
Makes sense, there will be only one case left then if the publish file and plate entity both get created... but the linking fails. But we can't re-run it just like that since the duplicate publish file on disk validation will fail.

And it won't show up in the conflicting publish list either as the finalize will run for that item and clear the status, since we are handling the exception. Thoughts?

Reply:
Not sure I fully follow, but ultimately, since you control the calling method, you should have access to all involved entities, as well as their linked files on disk, so if anything fails you should be able to clean it all up before exiting the process. Quite frankly, the parent class should probably be handling its own cleanup, if its not already...can't remember, in the case of failure.